src/handler.ts

```diff
@@ -53,7 +53,7 @@ export default class DataHandler {
     const { path } = this.target;
 
     if (value === undefined && args === undefined) {
-      Logger.getInstance().warn(`${path} emitted no data`, path);
+      Logger.getInstance().warn(LogMessageType.EventEmpty, { path });
```

Comment:
I wonder if it makes sense to have two different functions: `warn` which takes a raw string and `warnf` which takes a tokenized string and a variable list of parameters.

Reply:
I like the idea of a formatted `warn` or `error`: `warnf` and `errorf` respectively.  I started this PR with the question, "Where's the sprintf in JS?"  Line 56 above is actually the behavior of a message with a context.

```javascript
interface LogContext {
  rule?: string;
  source?: string;
  operator?: string;
  path?: string;
  property?: string;
  selector?: string;
  reason?: string;
}
```

The spirit of `LogContext` is to provide well-defined, critical information related to debugging.  Depending on scope, some of these members are unavailable, but at all times, at least one of them is likely to be available.  As such, the developer should populate as much of `LogContext` as possible.  `LogContext` also provided a way to fix inconsistencies in string messages and the fact that these error messages become longer (less manageable) the more context we add.

This PR added a `Logger.format` to re-use error messages with slight tweaks for clarity.  For example, `You forgot the )` and `You forgot the (` could be a template with substitution.  The `Logger.format` does not re-use the context though.  I was thinking that `format`'s job is to provide very specific information.  The `)` or `(` variable portion of `You forgot the` would need a member defined in context.  `token` could be added, but then I was concerned that `LogContext` fills up with other stuff and loses that narrow scope.

One solution could be we go with an `errorf(message: string, context: LogContext, ...substitutions?: string[])`.  One issue is that some of the code like `selector.ts` is so deep that it is very close to being completely unaware of context.  If we remove the context param, then the `errorf` feels a bit like `format`.

Open to alternatives or perspective from @TrevorFSmith.